core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -122,6 +124,61 @@ static double getSourceTablesSizeInBytes(PlanNode node, Lookup lookup, StatsProv
                 .sum();
     }
 
+    private static double getApproximateOutputSizeInBytes(PlanNode node, Context context)
+    {
+        return getApproximateOutputSizeInBytes(node, context.getLookup(), context.getStatsProvider(), context.getSymbolAllocator().getTypes());
+    }
+
+    @VisibleForTesting
+    static double getApproximateOutputSizeInBytes(PlanNode node, Lookup lookup, StatsProvider statsProvider, TypeProvider typeProvider)
+    {
+        if (hasExpandingNodes(node, lookup)) {
+            return NaN;
+        }
+
+        return getFirstKnownOutputSizeInBytes(node, lookup, statsProvider, typeProvider);
+    }
+
+    private static double getFirstKnownOutputSizeInBytes(PlanNode node, Lookup lookup, StatsProvider statsProvider, TypeProvider typeProvider)
+    {
+        return Stream.of(node)
+                .flatMap(planNode -> {
+                    if (planNode instanceof GroupReference) {
+                        return lookup.resolveGroup(node);
+                    }
+                    return Stream.of(planNode);
+                })
+                .mapToDouble(resolvedNode -> {
+                    PlanNodeStatsEstimate estimate = statsProvider.getStats(resolvedNode);
+                    if (!estimate.isOutputRowCountUnknown()) {
```

Comment:
do we need this check. I think `getOutputSizeInBytes` will return `NaN` if `outputRowCount` is `NaN`